### ISSUE TITLE
Allowed memory size of xxx bytes exhausted in current latest-container

### ISSUE THREAD

**Issue by pfaelzerchen** [nextcloud/docker#1413]
After upgrading my docker-setup to the current latest-container from the 21-branch some hours ago, I got the following error messages:
```
docker exec nextcloud su - www-data -s /bin/bash -c 'php -f /var/www/html/cron.php'

Fatal error: Allowed memory size of 2097152 bytes exhausted (tried to allocate 438272 bytes) in /var/www/html/3rdparty/composer/autoload_real.php on line 37
```

This happens basically with every php-command, e.g. occ to add the missing indices. Something seems wrong here with PHP Memory Limits. The following works fine:

```
docker exec nextcloud su - www-data -s /bin/bash -c 'php -d memory_limit=-1 -f /var/www/html/cron.php'
```

The PHP_MEMORY_LIMIT variable within the container seems to be set to 512. I found an older support-ticket regarding issues with this limit:
https://help.nextcloud.com/t/allowed-memory-size-of-xxx-bytes-exhausted/39371

So there seems an issue with the container's internal environment configuration.

**Comment by HomeITAdmin** [comment]
Just do add some information. This is also happening with the latest 19.0-apache and 20.0-apache and can be easily reproduced with a fresh new container from these images.

**Comment by ToeiRei** [comment]
Happens on latest PHP 7.4 and 8.0 too.

**Comment by SuperSandro2000** [comment]
The PHP_MEMORY_LIMIT variable introduced with https://github.com/nextcloud/docker/commit/061d860c0e711c3b4e9e08b2e67b34a4c33a1156 is not set when running ``docker exec``.

**Comment by pfaelzerchen** [comment]
> The PHP_MEMORY_LIMIT variable introduced with 061d860 is not set when running docker exec.

To be precise: It is not set, if you make the user switch to www-data. So the following will give you nothing

```
$ sudo docker exec nextcloud su - www-data -s /bin/bash -c 'env' | grep PHP_MEMORY_LIMIT
```

Executing this with root privileges in the container, will show a PHP_MEMORY_LIMIT

```bash
$ sudo docker exec nextcloud /bin/bash -c 'env' | grep PHP_MEMORY_LIMIT
PHP_MEMORY_LIMIT=512M
```

The given fixes work for me. However, using the -d switch of php (-d memory_limit=-1) does not work with 512, but with -1.

**Comment by dashohoxha** [comment]
I get the same error too:
```
Fatal error: Allowed memory size of 2097152 bytes exhausted (tried to allocate 36864 bytes) in /var/www/html/3rdparty/composer/autoload_static.php on line 3332
```

The latest image that works correctly for me is `nextcloud:20.0.6`.

> The given fixes work for me. However, using the -d switch of php (-d memory_limit=-1) does not work with 512, but with -1.

Don't you think that the right way to fix it is to release a new image that has the correct value?

**Comment by Nowaker** [comment]
The same happens on Arch Linux with `nextcloud` package. Creating the following file fixes the problem with `occ`:

```
# cat /etc/php7/conf.d/apcu.ini
[apcu]
apc.enable_cli=1
```

**Comment by ianhundere** [comment]
i'll chime in and ditto the fix of making `PHP_MEMORY_LIMIT` available as an env:
`sudo -u www-data PHP_MEMORY_LIMIT=512M php -f cron.php`

**Comment by Nowaker** [comment]
Adjusting the memory limit does not fix the issue. In my case 4 GB was not enough. People reported setting as high as 8 GB would not work for them. https://help.nextcloud.com/t/solved-occ-command-php-fatal-error-allowed-memory-size-of-xxx-bytes-exhausted/108521/11 Even 32 GB was mentioned but it wasn't clear to me the OP meant 32 GB is enough or not enough: https://help.nextcloud.com/t/solved-occ-command-php-fatal-error-allowed-memory-size-of-xxx-bytes-exhausted/108521/12

**Comment by ToeiRei** [comment]
@Nowaker do you use apcu? if yes, switch to redis as a memory cache.

**Comment by Killerhaschen** [comment]
For me it helped to set the user in in the [Docker exec](https://docs.docker.com/engine/reference/commandline/exec/) ommand:
docker exec -u 33 nextcloud_nextcloud_1 php -f /var/www/html/cron.php
User 33 is www-data.

**Comment by gjanssens** [comment]
I'm running into this as well. While I'm running from docker as well, I'm not using the docker image provided by Nextcloud. Instead my setup is based on two fedora containers, one with apache, the other with php-fpm.

The  default memory limit on my php installation is 1Gb.  With the memory limit I still run into occ out of memory issues. I have  tried to gradually increase the memory limit using -dmemory_limti=xyz. At 3Gb it's still breaking. With no memory limit (-1) occ will be killed by the OOM.

I have returned to nextcloud 20.0.11 until the real cause of this issue is found.

**Comment by PierrePetit** [comment]
I had the same issue today, on my Ubuntu server.

When I tried running my cron.php:
`docker exec nextcloud php -f /var/www/html/cron.php`
I had:
`PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes) in /var/www/html/apps/files_trashbin/lib/Helper.php on line 85`

I solved it simply by mapping a php.ini file in my docker-compose nextcloud's volumes like so:
`- ${CONTAINER_FILES_ROOT_PATH}/nextcloud/php.ini:/usr/local/etc/php/php.ini`
content:
`PHP_MEMORY_LIMIT=1024M`
I then restarted the nextcloud container and ran the cron.php once more. It then ran without the error. 
I think the process just needed a bit more RAM since I have more and more files on my nextcloud server (a few Tos).

Hope this helps someone

**Comment by dashohoxha** [comment]
In my case the problem is that I am extending the standard docker image by installing `systemd`, so `apache2` is called by `systemd` and in this case it cannot read the variables PHP_MEMORY_LIMIT and PHP_UPLOAD_LIMIT, which are defined in the root's environment.

In the default docker image apache2 is called with `exec apache2 ...`:
https://github.com/docker-library/php/blob/master/8.0/bullseye/apache/apache2-foreground#L40
So it inherits the environment of root and this is why it can access the variables PHP_MEMORY_LIMIT and PHP_UPLOAD_LIMIT.

My workaround is to modify `/usr/local/etc/php/conf.d/nextcloud.ini` directly with the correct values.

**Comment by Nowaker** [comment]
Any workarounds here fail to address the core problem: nextcloud, in certain cases, is a memory hog and even 8 GB RAM won't suffice. It's not even related to Docker, as explained in https://github.com/nextcloud/docker/issues/1413#issuecomment-787106975.

**Comment by dashohoxha** [comment]
@Nowaker Maybe you are right, but my problem was a different one, related to the PHP configuration. And my error message was exactly "Allowed memory size of xxx bytes exhausted ...". Probably we are talking about different issues.

**Comment by Nowaker** [comment]
@dashohoxha Your issue is technically off-topic because you introduced a problem (setting environment variables which have no chance of being passed to PHP) which you later resolved (by setting the memory limits correctly).

**Comment by dashohoxha** [comment]
@Nowaker Maybe you are right. Good luck with fixing your issue. Sorry for the confusion.

**Comment by pfaelzerchen** [comment]
> The same happens on Arch Linux with `nextcloud` package. Creating the following file fixes the problem with `occ`:
> 
> ```
> # cat /etc/php7/conf.d/apcu.ini
> [apcu]
> apc.enable_cli=1
> ```

I tried passing this option on the command line to test if it would work:

`sudo docker exec nextcloud su - www-data -s /bin/bash -c 'php --define apc.enable_cli=1 -f /var/www/html/cron.php'`

But the error remained.

**Comment by pfaelzerchen** [comment]
> do you use apcu? if yes, switch to redis as a memory cache.

I already have
`'memcache.local' => '\OC\Memcache\APCu',
'memcache.distributed' => '\OC\Memcache\Redis',`

After switching the local memcache to Redis, nothing changed. The error message remained.

**Comment by rhtenhove** [comment]
To bring back attention to the fix @dashohoxha mentioned. Using the default `nextcloud:fpm-alpine` image without any unsupported modification showed the same problem as mentioned in the issue title.

Overwriting the file `/usr/local/etc/php/conf.d/nextcloud.ini` with the values desired ensured these were used by both the Nextcloud instance itself (as seen under Settings - System), as well as `occ`, without any further modifications. This can be verified running for example `docker exec -u www-data nextcloud_container php -i | grep -i memory_limit`.

Using either the `configs` key for Docker Swarm, or a regular host mounted file, to overwrite that file prevents Nextcloud from trying (and failing) to get these keys from the environment.

Haven't tried this in other containers, such as the debian or apache images.

Is this a permanent solution? No. I'd like to have it use the values set in the environment. It does fix the problem for those struggling to get things working at all.

**Comment by dashohoxha** [comment]
> This can be verified running for example `docker exec -u www-data nextcloud_container php -i | grep -i memory_limit`

What happens if you call it like this:

```bash
docker exec \
    --env PHP_MEMORY_LIMIT=512M \
    --env PHP_UPLOAD_LIMIT=512M \
    -u www-data \
    nextcloud_container php -i \
  | grep -i memory_limit
```

If this works, then the problem is that the docker image **always** expects the environment variables `PHP_MEMORY_LIMIT` and `PHP_UPLOAD_LIMIT` to be defined by `--env` (or `-e`). If they are not defined, then the PHP configuration is broken.

One possible solution for this may be to fix the docker image so that it assumes some default values if these variables are not defined.

**Comment by rhtenhove** [comment]
It seems the omission of the `-p` flag for `su` causes this problem. This flag ensures the environment of the user running the `su` command (usually root in this case) is preserved. So the behavior described as problematic is actually as expected.

This preservation is taken care of in `/entrypoint.sh` (as well as in `/usr/local/etc/php-fpm.d/docker.conf` using `clear_env = no` for PHP-FPM), but the `run_as` function in that script is ignored when running a custom command. When running commands directly (using `exec`) this entrypoint script isn't even used.

So `docker exec nextcloud su -p www-data -s /bin/bash -c '/var/www/html/occ'` should run just fine. Or any other command for that matter, such as the `cron.php` script in the issue description.

Same for `docker exec -u www-data nextcloud /var/www/html/occ`.

These work because the image already has the correct environment variables set (these are set in the `Dockerfile`). These can be overwritten using `-e`, but that will only happen for that shell, and they will need to be preserved using `-p` if using `su` in a subsequent command.

If  `-p` isn't set, it will revert to the default which apparently isn't enough. Not sure what that is though, as it complains it has exhausted the 2M limit, but the default limit of PHP 8 is 128M.

Clearing the env:
```shell
$ docker exec nextcloud su - www-data -s /bin/sh -c 'php -i' | grep memory_limit
memory_limit => no value => no value
```

Keeping the env:
```shell
docker exec nextcloud su -p www-data -s /bin/sh -c 'php -i' | grep memory_limit
memory_limit => 512M => 512M
```

**Comment by pfaelzerchen** [comment]
@rhtenhove To me this seems to be the right solution to deal with the problem. So it basically was "a user problem".

I searched for this in the README for the docker image. There I found a slightly different command:
`docker exec --user www-data nextcloud php occ`
which also works for cron.php
`docker exec --user www-data nextcloud php cron.php`

Both commands work without any trouble. So Nextcloud has a proper documentation on this.
